walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -36,50 +39,171 @@ public int compare(Event a, Event b) {
+        j++;
+      }
+      prevTime = (Integer) changeLog.get(i).getKey();
+      currAttendance += (Integer) changeLog.get(i).getValue();
+    }
+
+    // If there are no meeting times with at least one optional attendee, just return the good
+    // windows.
+    return optimalMeetingTimes.isEmpty() ? goodWindows : optimalMeetingTimes;
   }
 
-  private Collection<TimeRange> getMeetingTimes(
-      Collection<Event> eventsCollection,
-      MeetingRequest request,
-      boolean includeOptionalAttendees) {
-    HashSet<String> attendees = new HashSet<String>(request.getAttendees());
-    if (includeOptionalAttendees) {
-      attendees.addAll(request.getOptionalAttendees());
+  /**
+   * Returns a mapping of optional attendees to their free times.
+   *
+   * @param events all events to be considered
+   * @param optionalAttendees everyone who needs to attend this meeting
+   * @param minTime minimum length of time for meeting
+   */
+  private HashMap<String, ArrayList<TimeRange>> getOptionalAttendeesFreeTimes(
+      Collection<Event> events, HashSet<String> optionalAttendees, long minTime) {
+    HashMap<String, ArrayList<TimeRange>> times = new HashMap<String, ArrayList<TimeRange>>();
+    for (String attendee : optionalAttendees) {
+      HashSet<String> attendeeSet = new HashSet<String>();
+      attendeeSet.add(attendee);
+
+      // Find all possible meeting times for just this one attendee. Must do this to deal with
+      // double bookings.
+      times.put(attendee, getMandatoryAttendeesMeetingTimes(events, minTime, attendeeSet));
     }
-    ArrayList<Event> events = getRelevantEvents(attendees, new ArrayList<Event>(eventsCollection));
-    List<TimeRange> possibleMeetingTimes = new ArrayList<TimeRange>();
+    return times;
+  }
+
+  /**
+   * Returns a change log of the number of available optional attendees over time. First part of a
+   * sweep-line algorithm.
+   *
+   * @param optionalAttendeesFreeTimes mapping of all attendees to their free times
+   */
+  private ArrayList<Map.Entry<Integer, Integer>> getChanges(
+      HashMap<String, ArrayList<TimeRange>> optionalAttendeesFreeTimes) {
+    TreeMap<Integer, Integer> changes = new TreeMap<Integer, Integer>();
+    for (Map.Entry e : optionalAttendeesFreeTimes.entrySet()) {
+      for (TimeRange time : (ArrayList<TimeRange>) e.getValue()) {
+        changes.put(time.start(), changes.getOrDefault(time.start(), 0) + 1);
+        changes.put(time.end(), changes.getOrDefault(time.end(), 0) - 1);
+      }
+    }
+    return new ArrayList<Map.Entry<Integer, Integer>>(changes.entrySet());
+  }
+
+  /**
+   * Updates optimalMeetingTimes based on current meeting time's attendance.
+   *
+   * @param time current meeting time range
+   * @param bestAttendance best attendance of optional attendees seen so far
+   * @param currAttendance attendance optional attendees of current meeting
+   * @param optionalAttendeesFreeTimes all attendees and their free times
```

Comment:
This parameter name does not match.

Also, parameters that are edited by the method (in this case `optimalMeetingTimes`) are traditionally put at the end of the parameter list.